### PR TITLE
[Optimizer] Add Graph::to_dag()

### PR DESCRIPTION
**Pull request by xumingkuan**
This PR adds a function `Graph::to_dag()`.

We can then use `graph.to_dag()->hash(ctx)` to do random testing and get a fingerprint of `graph`, where `ctx` stores the necessary contexts for random testing. In this way, for any two `Graph`s with the same functionality, say `graph1` and `graph2`, the values of `graph1.to_dag()->hash(ctx)` and `graph2.to_dag()->hash(ctx)` must be:
- either the same
- or due to floating-point errors, 
  - `graph1.to_dag()->hash(ctx)` is in `graph2.to_dag()->other_hash_values()`,
  - or `graph2.to_dag()->hash(ctx)` is in `graph1.to_dag()->other_hash_values()`.
Note that `DAG::other_hash_values()` is cached after calling `DAG::hash(Context *)`. For now, `DAG::other_hash_values()` is actually a `std::vector` with only 1 element which is `DAG::hash(Context *)` plus one.

### CODE REVIEW

**Comment by jiazhihao** [comment]
@xumingkuan @zikun-li Do you think we can merge this PR? Emily's project requires this function to manage subgraphs.

**Comment by zikun-li** [comment]
> @xumingkuan @zikun-li Do you think we can merge this PR? Emily's project requires this function to manage subgraphs.

I've leave a comment for Mingkuan, I think it's okay to merge this PR if he take that into consideration and thinks the implementation works.

**Comment by jiazhihao** [comment]
I didn't see the comment. It would be great if we can merge this PR now since the PR currently blocks Emily's REUSE project. Thanks!

**Comment by zikun-li** [comment]
Okay I think we can merge it for now. 
I left the following comment 4 days ago:
>Does this depends on some assumptions of guid? Just in case. guid is not guaranteed to indicate the topology (or any other) order essentially. Actually its behavior is pretty arbitrary. For example if we use the equivalence rz(p0) + rz(p1) = rz(add(p0 + p1)) to merge two rz, the parameter gate are unchanged after the transformation, however, if we eliminate the add and merge the two input_param together, there will be a new input_param who will have a new guid, and that guid will be larger than all guid used in the context.

**Comment by jiazhihao** [comment]
Hmmm, I think this is a valid concern. Given that no other functions depend on `Graph::to_dag()`, merging this PR shouldn't break any existing functionalities. I have merged the PR but we should figure out whether it makes sense to use guid to rank `Op`s

File: src/quartz/tasograph/tasograph.cpp [hunk 1]
```diff
@@ -162,6 +162,117 @@ Graph::Graph(const Graph &graph) {
...
+  for (const auto &it: outEdges) {
+    if (it.first.ptr->tp == GateType::input_param && op_2_param_idx.count(it.first) == 0) {
+      int idx = num_input_params++;
+      op_2_param_idx[it.first] = std::vector<int>(1, idx);
+      guid_and_param_index.emplace_back(it.first.guid, idx);
+      gates.push(it.first);
+    }
+  }
+  // Sort input parameter index by guid.
+  // This is the order originally in DAG.
```
[the remark this answers is not in the document]
Comment (zikun-li) [reply src/quartz/tasograph/tasograph.cpp:185]: It seems I didn't click the `review changes` on Monday so that others cannot see this....

**Comment by xumingkuan** [comment]
The (priority) queue is for topological sorting. By comparing by guid, I am guaranteeing that if a `DAG` is transformed to a `Gragh` and immediately transformed back, the order of the gates should be the same, that is, it will still have the canonical order.

For other cases, like @zikun-li said, the guid order can be arbitrary. My implementation should still guarantee that it is a valid topological order, but seems not guaranteeing canonical order.

**Pull request merged**